### ISSUE TITLE
feature request: Async DDPG (D3PG and D4PG)

### ISSUE THREAD

**Issue by M00NSH0T** [keras-rl/keras-rl#185]
I'd love to see an asynchronous version of DDPG on here. Would anyone be able to help me with it?

Here are quick thoughts:

A3C seems to be king of the hill at the moment, but DDPG has some clear advantages in terms of generalizing better, and ultimately is very similar in its implementation to A2C.  It just doesn't multithread, so A3C can win with greater musclepower. But if DDPG had a readily available and easy to use async version as well in Keras-RL, I think it would get a lot of widespread love.

However, the A3C implementation in the experimental branch never quite got done, so it's going to take some effort to get that up and running, then rework it into an Async DDPG. I think I know the problem though.
In the [paper,](https://arxiv.org/pdf/1602.01783.pdf) Deepmind states that all the actor / critic layers should be shared (i.e. with common weights) except the last one, but in this A3C implementation, the actors and critics seem to have independent networks. I suspect that's the reason there are issue threads here regarding its performance. The shared weights seem to be a fundamental mechanism by which these asynchronous threads work together toward the common goal, so with the actors and critics being independent networks, it seems like memory recall with a common pool of memories would be the only thing binding them together. I could be wrong though. 

Next, one of the main differences between A3C and an async DDPG is the fact that DDPG takes action as an additional input with state in the critic network.  However, I'm not entirely sure if you also need to share the core layers of the actor and critic networks in an async DDPG implementation the way you're supposed to with A3C, and if you are, how you would do that. I saw one implementation (which I'll have to find again) where the output of the critic was simply multiplied by the action input right before the last layer, but that doesn't totally make sense to me. In that implementation though the actors and critics did share weights up till that point. I do see how maybe you could have two neural networks (one built just on state, and a simpler one on action with maybe just one layer) that merge together, and then have maybe 1-2 additional hidden layers before the output. I'm not sure if that would totally work, if there's a better way to do it, or if that's just totally not necessary.

I did find another A3C implementation built with Keras that has worked well for me. [Check this out.](https://github.com/rlcode/reinforcement-learning/blob/master/2-cartpole/5-a3c/cartpole_a3c.py). You can see that the actor / critic layers are shared, though they did diverge for more than just the last layer. In my version, I kept it like in the paper and all layers were shared except the output (critic just has the Q value as output, and actor gets a softmax vector). For async DDPG, I think it would be a similar network setup, just with a more complex critic network that also works the action in somewhere. I just don't know how many layers should be shared with the actor(s).

**Comment by M00NSH0T** [comment]
been doing some more reading on this. Actually what we would want is specifically an async version of D4PG.  This looks the gold standard these days.
https://arxiv.org/pdf/1801.00690.pdf

Here's the paper outlining it:
https://openreview.net/pdf?id=SyZipzbCb

Essentially, predicting scaler Q-values is a thing of the past. We really want to predict distributions instead of average/expected returns. D4PG would let us both do asyncronous training, and also use the distributional form of the bellman equation, which has a ton of advantages.

**Comment by M00NSH0T** [comment]
sorry closed by accident...

**Comment by nskplta** [comment]
Hi @M00NSH0T , could no make [a3c_pendulum.py](https://github.com/keras-rl/keras-rl/blob/experimental/examples/a3c_pendulum.py) work, lots of updates and errors for Keras 2.0, however your other suggestion(https://github.com/rlcode/reinforcement-learning/blob/master/2-cartpole/5-a3c/cartpole_a3c.py) works. Where can I find your version? 
> In my version, I kept it like in the paper and all layers were shared except the output (critic just has the Q value as output, and actor gets a softmax vector)

Any implementation of D4PG?

Thanks a lot!

**Comment by svenuthe** [comment]
Can someone help to implement this?
